lib/spack/spack/solver/concretize.lp

```diff
@@ -555,7 +555,8 @@ variant(Package, Variant) :- variant_condition(ID, Package, Variant),
 attr("variant_propagate", Package, Variant, Value, Source) :-
   attr("node", Package),
   depends_on(Parent, Package),
-  attr("variant_propagate", Parent, Variant, Value, Source),
+  attr("variant_propagate", Parent, _, _, Source),
+  attr("variant_value", Source, Variant, Value),
```

Comment:
Hmm, this seems wrong - wouldn't it propagate all the variants from `Source`? Asking since, if we don't match on the middle arguments of:
```
attr("variant_propagate", Parent, _, _, Source),
```
then 
```
attr("variant_value", Source, Variant, Value),
```
will not be related to the arguments in `variant_propagate`

Reply:
I confirm this is wrong. With this diff:
```diff
diff --git a/var/spack/repos/builtin/packages/zlib/package.py b/var/spack/repos/builtin/packages/zlib/package.py
index e9425f93bf..d37923f018 100644
--- a/var/spack/repos/builtin/packages/zlib/package.py
+++ b/var/spack/repos/builtin/packages/zlib/package.py
@@ -48,7 +48,7 @@ class Zlib(MakefilePackage, Package):
     build_system("makefile", conditional("generic", when="platform=windows"), default="makefile")
 
     variant("pic", default=True, description="Produce position-independent code (for shared libs)")
-    variant("shared", default=True, description="Enables the build of shared libraries.")
+    variant("shared", default=False, description="Enables the build of shared libraries.")
     variant("optimize", default=True, description="Enable -O2 for a more optimized lib")
 
     conflicts("build_system=makefile", when="platform=windows")
```
to ensure `zlib` default for `shared` is `False`, I get:
```console
$ spack spec hdf5~mpi
Input spec
--------------------------------
 -   hdf5~mpi

Concretized
--------------------------------
 -   hdf5@1.14.1-2%gcc@9.4.0~cxx~fortran~hl~ipo~java~map~mpi+shared~szip~threadsafe+tools api=default build_system=cmake build_type=Release generator=make arch=linux-ubuntu20.04-icelake
 -       ^cmake@3.26.3%gcc@9.4.0~doc+ncurses+ownlibs build_system=generic build_type=Release arch=linux-ubuntu20.04-icelake
 -           ^curl@8.1.2%gcc@9.4.0~gssapi~ldap~libidn2~librtmp~libssh~libssh2+nghttp2 build_system=autotools libs=shared,static tls=openssl arch=linux-ubuntu20.04-icelake
 -               ^nghttp2@1.52.0%gcc@9.4.0 build_system=autotools arch=linux-ubuntu20.04-icelake
 -               ^openssl@3.1.2%gcc@9.4.0~docs~shared build_system=generic certs=mozilla arch=linux-ubuntu20.04-icelake
 -                   ^ca-certificates-mozilla@2023-05-30%gcc@9.4.0 build_system=generic arch=linux-ubuntu20.04-icelake
 -                   ^perl@5.38.0%gcc@9.4.0+cpanm+opcode+open+shared+threads build_system=generic arch=linux-ubuntu20.04-icelake
 -                       ^berkeley-db@18.1.40%gcc@9.4.0+cxx~docs+stl build_system=autotools patches=26090f4,b231fcc arch=linux-ubuntu20.04-icelake
 -                       ^bzip2@1.0.8%gcc@9.4.0~debug~pic+shared build_system=generic arch=linux-ubuntu20.04-icelake
 -                           ^diffutils@3.9%gcc@9.4.0 build_system=autotools arch=linux-ubuntu20.04-icelake
 -                               ^libiconv@1.17%gcc@9.4.0 build_system=autotools libs=shared,static arch=linux-ubuntu20.04-icelake
 -                       ^gdbm@1.23%gcc@9.4.0 build_system=autotools arch=linux-ubuntu20.04-icelake
 -                           ^readline@8.2%gcc@9.4.0 build_system=autotools patches=bbf97f1 arch=linux-ubuntu20.04-icelake
 -           ^ncurses@6.4%gcc@9.4.0~symlinks+termlib abi=none build_system=autotools arch=linux-ubuntu20.04-icelake
 -       ^gmake@4.4.1%gcc@9.4.0~guile build_system=autotools arch=linux-ubuntu20.04-icelake
 -       ^pkgconf@1.9.5%gcc@9.4.0 build_system=autotools arch=linux-ubuntu20.04-icelake
 -       ^zlib@1.2.13%gcc@9.4.0+optimize+pic~shared build_system=makefile arch=linux-ubuntu20.04-icelake
```
when not using propagation and:
```console
$ spack spec hdf5~~mpi
Input spec
--------------------------------
 -   hdf5~~mpi

Concretized
--------------------------------
 -   hdf5@1.14.1-2%gcc@9.4.0~cxx~fortran~hl~ipo~java~map~mpi+shared~szip~threadsafe+tools api=default build_system=cmake build_type=Release generator=make arch=linux-ubuntu20.04-icelake
 -       ^cmake@3.26.3%gcc@9.4.0~doc+ncurses+ownlibs build_system=generic build_type=Release arch=linux-ubuntu20.04-icelake
 -           ^curl@8.1.2%gcc@9.4.0~gssapi~ldap~libidn2~librtmp~libssh~libssh2+nghttp2 build_system=autotools libs=shared,static tls=mbedtls arch=linux-ubuntu20.04-icelake
 -               ^mbedtls@3.3.0%gcc@9.4.0+pic build_system=makefile build_type=Release libs=static arch=linux-ubuntu20.04-icelake
 -               ^nghttp2@1.52.0%gcc@9.4.0 build_system=autotools arch=linux-ubuntu20.04-icelake
 -           ^ncurses@6.4%gcc@9.4.0~symlinks+termlib abi=none build_system=autotools arch=linux-ubuntu20.04-icelake
 -       ^gmake@4.4.1%gcc@9.4.0~guile build_system=autotools arch=linux-ubuntu20.04-icelake
 -       ^pkgconf@1.9.5%gcc@9.4.0 build_system=autotools arch=linux-ubuntu20.04-icelake
 -       ^zlib@1.2.13%gcc@9.4.0+optimize+pic+shared build_system=makefile arch=linux-ubuntu20.04-icelake
```
when using propagation, i.e. all the variants from `hdf5` are propagated.

We need to come up with more tests for cases of interest to add to the test suite.